kolibri/core/webpack/hooks.py

```diff
@@ -0,0 +1,290 @@
+
+        An auto-generated path to where the build-time files are stored,
+        containing information about the built bundles.
+        """
+        return os.path.join(
+            self.build_path,
+            "{plugin}_stats.json".format(plugin=self.unique_slug)
+        )
+
+    @property
+    def _module_file_path(self):
+        """
+        Returns the path of the class inheriting this classmethod.
+        """
+        return os.path.join(*self.__module__.split(".")[:-1])
+
+    def render_to_html(self, extension=None):
+        """
+        This function tags a bundle of file chunks and generates the appropriate
+        script tags for them, be they JS or CSS files.
+
+        :param bundle_data: The data returned from
+        :return: HTML of script tags for insertion into a page.
+        """
+        tags = []
+        js_tag = '<script type="text/javascript" src="{url}"></script>'
+        css_tag = '<link type="text/css" href="{url}" rel="stylesheet"/>'
+        for chunk in self.bundle_filtered(extension=extension):
+            if chunk['name'].endswith('.js'):
+                tags.append(js_tag.format(url=render_as_url(chunk)))
+            elif chunk['name'].endswith('.css'):
+                tags.append(css_tag.format(url=render_as_url(chunk)))
+        return mark_safe('\n'.join(tags))
+
+
+class FrontEndAssetHook(WebpackBundleHook):
+    """
+    An abstract class for all assets destined for the default front-end. You
+    probably want to use FrontEndSyncHook or FrontEndASyncHook to be explicit.
+
+    Everything inheriting from this hook will automatically be included in the
+    front-end.
+    """
+
+    class Meta:
+        abstract = True
+
+
+class FrontEndSyncHook(FrontEndAssetHook):
+    """
+    Define something that should be included for sync'ed purposes. Assets will
+    always be loaded.
+    """
+
+    class Meta:
+        abstract = True
+
+
+class FrontEndCoreHook(FrontEndSyncHook):
+    """
+    A hook that asserts its only applied once, namely to load the core
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(FrontEndCoreHook, self).__init__(*args, **kwargs)
+        assert len(list(self.registered_hooks)) <= 1, "Only one core asset allowed"
+
+    @property
+    @hooks.registered_method
+    def webpack_bundle_data(self):
+        dct = super(FrontEndCoreHook, self).webpack_bundle_data
+        dct['core'] = True
+        dct['external'] = True
+        return dct
+
+    class Meta:
+        abstract = True
+
+
+class FrontEndASyncHook(FrontEndAssetHook):
+    """
+    Define something that should be included for sync'ed purposes.
+    """
+
+    # : A list of events to listen to
+    events = {}
+
+    # : A list of events to load the asset once
+    once = {}
+
+    @property
+    @hooks.registered_method
+    def webpack_bundle_data(self):
+        dct = super(FrontEndASyncHook, self).webpack_bundle_data
+        dct['events'] = self.events
```

Comment:
I would that this were not so - I had tried to put all this events business just directly in the Javascript (where it made the most sense to have it), and then extract the stuff we needed for async loading at Javascript build time. Unfortunately, this was too brittle.

By having it here, we at least have the benefit of only defining it in one place, even though it is still not great to have what is essentially JS configuration in Python.